packages/langium/src/generator/generator-node.ts

```diff
@@ -4,31 +4,83 @@
  * terms of the MIT License, which is available in the project root.
  ******************************************************************************/
 
-const EOL = (typeof process === 'undefined') ? '\n' : (process.platform === 'win32') ? '\r\n' : '\n';
+import { processGeneratorNode } from './node-processor';
 
-export type GeneratorNode = CompositeGeneratorNode | IndentNode | NewLineNode | string;
+export const EOL = (typeof process === 'undefined') ? '\n' : (process.platform === 'win32') ? '\r\n' : '\n';
+
+export type Generated = GeneratorNode | string | undefined;
+export type GeneratorNode = CompositeGeneratorNode | IndentNode | NewLineNode;
+
+export function isGeneratorNode(node: unknown): node is CompositeGeneratorNode | IndentNode | NewLineNode {
+    return node instanceof CompositeGeneratorNode
+        || node instanceof IndentNode
+        || node instanceof NewLineNode;
+}
+
+export function isNewLineNode(node: unknown): node is NewLineNode {
+    return node instanceof NewLineNode;
+}
+
+export function toString(object: unknown): string {
+    if (isGeneratorNode(object))
+        return processGeneratorNode(object);
+    else
+        return String(object);
+}
 
 export class CompositeGeneratorNode {
 
-    readonly contents: GeneratorNode[] = [];
+    readonly contents: Array<(GeneratorNode | string)> = [];
 
-    constructor(...contents: GeneratorNode[]) {
+    constructor(...contents: Generated[]) {
         this.append(...contents);
     }
 
-    append(...args: Array<GeneratorNode | ((node: CompositeGeneratorNode) => void)>): CompositeGeneratorNode {
+    isEmpty(): boolean {
+        return this.contents.length === 0;
+    }
+
+    append(...args: Array<Generated | ((node: CompositeGeneratorNode) => void)>): this {
         for (const arg of args) {
             if (typeof arg === 'function') {
                 arg(this);
-            } else {
+            } else if (arg) {
                 this.contents.push(arg);
             }
         }
         return this;
     }
 
-    indent(func?: (indentNode: IndentNode) => void): CompositeGeneratorNode {
-        const node = new IndentNode();
+    appendIf(condition: boolean, args: Generated | ((node: CompositeGeneratorNode) => void)): this {
+        return condition ? this.append(args) : this;
+    }
+
+    newLine(): this {
+        return this.appendNewLine();
+    }
+
+    appendNewLine(): this {
```

Comment:
Why two methods that do the same?

Reply:
It's syntactic sugar. I wasn't sure which syntax to propose, I actually prefer the longer one, but that's probably my personal taste.
Use cases are
 * `node.append(member.name).append(': {').appendNewLine()...`
 * ```
   return toNode`
        #include <iostream>
        #include <map>
        #include <string>

        class ${ctx.statemachine.name};

        ${generateStateClass(ctx)}

        ${generateStatemachineClass(ctx)}

        ${forEachNL(ctx.statemachine.states, state => generateStateDeclaration(ctx, state))}
        ${forEachNL(ctx.statemachine.states, state => generateStateDefinition(ctx, state))}

        typedef void (${ctx.statemachine.name}::*Event)();

        ${generateMain(ctx)}
   `.appendNewline();        // one way of enforcing a terminating line break.
   ```

Reply:
Removed duplicate methods, will add js doc now.